export_schema.py

```diff
@@ -0,0 +1,279 @@
+    sparse_coo = auto()
+    sparse_csr = auto()
+    sparse_csc = auto()
+    sparse_bsr = auto()
+    sparse_bsc = auto()
+    _mkldnn = auto()
+
+
+# Copied from torch/_C/__init__.pyi.in
+class MemoryFormat(Enum):
+    # Defined in torch/csrc/utils/tensor_memoryformats.cpp
+    contiguous_format = auto()
+    channels_last = auto()
+    channels_last_3d = auto()
+
+    # !!! This is annoying! preserve_format cannot be used as a tensor propery, as it can only used as an argument for operators
+    preserve_format = auto()
+
+
+# Copied from torch/_C/__init__.pyi.in
+@dataclass
+class Device:
+    # Defined in torch/csrc/Device.cpp
+    type: str
+    index: int
+
+
+SymInt = Union[
+    int,
+    str
+]
+
+Scalar = Union[int, float, bool]
+
+
+# ArgumentType = Union[
+
+#     Tensor
+#     Tensor?
+#     Tensor[]
+
+#     Scalar
+#     Scalar?
+
+#     ScalarType
+#     ScalarType?
+
+#     bool
+#     bool[3]
+#     bool?
+
+#     int
+#     int?
+#     int[]
+#     int[]?
+
+#     float
+#     float[]?
+
+#     SymInt
+#     SymInt[]
+#     SymInt[]?
+
+#     MemoryFormat?
+
+#     Layout
+#     Layout?
+
+#     Device
+#     Device?
+# ]
+
+################################################################################
+# Following section is the defining the schema of serializing a concrete tensor
+
+# TensorMeta is a decription of a tensor, without the actual data (,effectively maps to FakeTensor)
+# TensorMeta has multliple uses
+#   1. Represent the property of a concrete tensor backed by a storage
+#     - This is used in the serialization of a concrete tensor, e.g. model weight
+#     - In this case, sizes and strides must be concrete ints, and cannot be symbolic
+#     - stride and storage_offset have to used to correctly reconstruct the tensor from the storage
+#   2. Represent the property of a virtual tensor (see IValue below)
+#     - In this case, sizes and strides can be either concrete ints or symbolic ints.
+#     - device/strides/storage_offset/layout/memory_format are tied to pytorch's implementation.
+#       These are faithful capture of tensor's detail in pytorch's executions during tracing
+#       However, it's up to downstream system on how to utilized these fields
+#       In another word, these feilds are suggestive, rather than mandatory.
+@dataclass
+class TensorMeta:
+    dtype: ScalarType
+    sizes: List[SymInt]
+
+    # needed for training
+    requires_grad: bool
+
+    # !!! see description above, there are subtle difference on how these fields should be interpreted
+    device: Device
+    strides: List[SymInt]
+    storage_offset: int
+    layout: Layout
+    memory_format: MemoryFormat
+
+
+@dataclass
+class Buffer:
+    buffer: bytes
+
+
+@dataclass
+class ExternalBuffer:
+    # !!! Example from onnx
+    #   // Data can be stored inside the protobuf file using type-specific fields or raw_data.
+    #   // Alternatively, raw bytes data can be stored in an external file, using the external_data field.
```

Comment:
Noob here: can you share an example of `ExternalBuffer`?

Reply:
ExternalBuffer {
   location = "./tensor.dta",
   offset = "0",
   length = "2147483648", # 2GB
   checksum = "68e9da68cbeb1288b904022d237c32e88e0372fd"
}
}

Reply:
ok so basically this is not in memory?